### ISSUE TITLE
nthOr headOr

### ISSUE THREAD

**Issue by jethrolarson** [ramda/ramda#2111]
```
nthOr :: a -> Int -> [a] -> a
headOr :: a -> [a] -> a
```
Gives safe way of extracting a value from a list that can be empty, similar to `propOr`

**Comment by CrossEye** [comment]
An interesting alternative to introducing `Maybe`  into Ramda proper.

I'd have to think about it a bit.  I would worry that such things would proliferate and obviously they are easy to compose on ones own.  But they are convenient.

**Comment by jethrolarson** [comment]
Yeah, as much as I like sanctuary, this tactic that ramda has adopted of requiring defaults for safe functions is a decent idea that I think is more approachable than Maybes.

Obviously it has it's own pitfalls:
```js
const getStuff = propOr([], 'stuff')
// surprise impurity due to the mutability of arrays
```

**Comment by jethrolarson** [comment]
For the record, I'm still in support of ramda getting `Maybe`

**Comment by adispring** [comment]
We can compose some ramda 'get' API with `R.defaultTo`:

```js
const nthOr = R.curry((defaultVal, list) => R.compose(
  R.defaultTo(defaultVal),
  R.nth
)(list));
```

It's convenient, but too many api will need to write xxOr version. So I also support import `data.Maybe`, or compose APIs to do that by ourselves.

**Comment by Jacse** [comment]
@adispring alternatively we should create a function called R.or with a type signature of a -> (b -> a) -> b -> a

That way we could do: R.or([], R.prop('children')) and thus the user could wrap all existing functions easily without having to compose. Not sure if regular composition is preferable though.

On mobile so sorry for the poor formatting, and apologies for the probably wrong type signature, I hope you catch my drift.

**Comment by jethrolarson** [comment]
There's already R.or and R.either with different signatures

**Comment by buzzdecafe** [comment]
> For the record, I'm still in support of ramda getting `Maybe`

... bumping this ...

**Comment by adispring** [comment]
@Jacse ,
```
const default = R.curry((defaultVal, func) => R.compose(
  R.defaultTo(defaultVal),
  func,
));
```

**Comment by CrossEye** [comment]
@adispring: how about `addDefault` in parallel with `addIndex`?

**Comment by adispring** [comment]
ok, I'll make a pr with api named `addDefault`.

**Comment by davidchambers** [comment]
@adispring, you seem to be ignoring the existence of `undefined`.

```javascript
> S.maybe('EMPTY', S.toString, S.head([]))
'EMPTY'

> S.maybe('EMPTY', S.toString, S.head([undefined]))
'undefined'

> R.defaultTo('EMPTY', R.head([]))
'EMPTY'

> R.defaultTo('EMPTY', R.head([undefined]))
'EMPTY'
```

The problem is that `R.head` does not preserve an important piece of information: whether or not it was applied to the empty array. Having lost this information, we cannot regain it. An `=== undefined` check gives us the right answer in most cases, but Ramda is a foundational library which should provide the right answer in *all* cases. That's my view, anyway. :)

**Comment by CrossEye** [comment]
Is this the expected behavior with Sanctuary?

```js
let arr = [];
arr[3] = undefined;

arr; //=> [ , , , undefined ]

S.maybe('EMPTY', S.toString, S.head(arr)); //=> 'undefined'
```

I find the combination of `undefined` and sparse arrays extremely disconcerting.

**Comment by adispring** [comment]
It's a problem to determine whether the returned `undefined` is an original one, or a calculated one.

**Comment by davidchambers** [comment]
@CrossEye, `Just(undefined)` is the only reasonable result of evaluating `S.head([,,,undefined])`, is it not? `Nothing` would indicate that the input was `[]`.

**Comment by CrossEye** [comment]
> `Nothing` would indicate that the input was `[]`.

That's not at all clear to me.  `Nothing` should indicate that the item supplied has no `head`.  But what does that mean with a sparse array without a first element?  Should `'0' in list; //=> false` imply that `head(list); //=> Nothing`  It's a reasonable alternative.  

I understand that since JS can return `undefined` as a value, the algebraic types need to be able to wrap it.  But it becomes uncomfortable when sparse arrays are included.  That's one good reason that in Ramda we've always punted on them.  But that's probably not a luxury that Sanctuary can afford.

**Comment by jethrolarson** [comment]
The head of a sparse array, that's kinda zen. Like one hand clapping or
something
